test/parallel/test-assert-deep.js

```diff
@@ -466,4 +466,11 @@ assertOnlyDeepEqual(
   assertDeepAndStrictEqual(m3, m4);
 }
 
+// Handle NaN
+assert.throws(() => { assert.deepEqual(NaN, NaN); }, assert.AssertionError);
```

Comment:
This is not intuitive. I'd assume that
![image](https://user-images.githubusercontent.com/96947/29738113-c66fbb3a-89e8-11e7-9402-c2e91e282abe.png)
(deepStrictEqual should be more strict than deepStrictEqual).
What's the rationale for not allowing `assert.deepEqual(NaN, NaN)`?

Reply:
That is the bad part about the wording with `strict`. The strict stands for more exact in this case. `deepEqual` is actually more like a `looksSimilar` and it actually fails at seeing the similarity in this case even though it exists. I can add that check to deepEqual as well but I think it would be weird to have it there (even though I would call deepEqual broken anyway - we actually also use a own lint rule to prevent usage of it!).

Reply:
Actually that is IMHO a great idea, create a better named superset assertion. That would also make this change semver-minor.

`sameValues` might be a better name (that way neither the word "strict" nor "deep" are necessary)

Reply:
I think this is absolutely not the right thing to do. And as I tried to explain earlier `deepStrictEqual` is not strictly a superset of `deepEqual` because it was just a poor naming from the beginning on.  Now we have to live with that naming. Most people do for example not know that `NaN` is not equal to itself when compared with `strict equality`. And they also do not know about the other existing equality comparisons in JS. They mostly do not even know how `abstract equality` is defined - at least that is what my experience tells me.
The comparisons in general are well described [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness). Adding a new API would not help anyone and would instead create more confusion.

Reply:
I agree that `assert.deepStrictEqual(NaN,NaN)` makes sense intuitively. 
But ECMAScript `abstract equal` is "broken" in this sense (`(NaN == NaN) === false`) and `strict equal` is broken (`(NaN === NaN) === false`) is the exact same way.
So IMHO we should be consistent with the standard if we keep using the term `equal` in our end-point names.

That's why I think breaking away for legacy with a `same` endpoint might be a good idea.

Reply:
The deepEqual and deepStrictEqual functions have much more then just the "basic" comparison and therefore do not correspond to these properly. I would love if that would be the only distinction because in that case `deepEqual` would not be so badly broken but that is not the case and it will not change either.
Adding another random name makes the issue not less an issue but just added more problems.